### PR TITLE
Refactor SqlTranslatingExpressionVisitor

### PR DESCRIPTION
**Pull request by tuespetre**
This commit adds the `OuterPropertyExpression`, which is a dedicated expression to represent expressions bound to parent query models. This expression allows us to defer the decision of whether to bind directly to the parent query (i.e. an `AliasExpression`) or to bind via an outer parameter. There are several improved generated SQL queries in this PR as a result.

To facilitate this change, other changes take place:

- Subquery models are now visited only once, meaning no more unnecessary uniquification (things like `FROM Orders AS o4` instead of `FROM Orders AS o`)
- The `_bindParentQueries` field/parameter to `RelationalQueryModelVisitor` and `SqlTranslatingExpressionVisitor` is now gone, instead being driven by the context of the query model visitor and its ancestors via `CanBindOuterProperties`, `CanBindOuterParameters`, and `RequiresOuterParameterInjection`
- The `_targetSelectExpression` field/parameter to `SqlTranslatingExpressionVisitor` is now gone, being replaced by a `_mutateProjections` field/parameter. This helps to make it clearer what is being done every time and place the visitor is used (also, there were cases where the `_targetSelectExpression` comparison bit would hinder things from binding correctly.)

This PR is split from (and integral to) #7650 and #7543, and this PR includes/depends on the changes that make up #7619.

### CODE REVIEW

**Comment by tuespetre** [comment]
List of queries with improved generated SQL (ignoring those that just changed in alias names due to the uniquification):

- https://github.com/aspnet/EntityFramework/pull/7663/files#diff-4685ea6da86b37b446cabc9dbe924057R1658
- https://github.com/aspnet/EntityFramework/pull/7663/files#diff-4685ea6da86b37b446cabc9dbe924057R2438
- https://github.com/aspnet/EntityFramework/pull/7663/files#diff-2c68636402e93e86a3d0108bbd4ba1ddL2074
- https://github.com/aspnet/EntityFramework/pull/7663/files#diff-e8ff8dabc15c4214f1f4885c86b23ce2R190
- https://github.com/aspnet/EntityFramework/pull/7663/files#diff-e8ff8dabc15c4214f1f4885c86b23ce2R448
- https://github.com/aspnet/EntityFramework/pull/7663/files#diff-e8ff8dabc15c4214f1f4885c86b23ce2R1017
- https://github.com/aspnet/EntityFramework/pull/7663/files#diff-2602188597986ba3e53c5b56162786d2L265
- https://github.com/aspnet/EntityFramework/pull/7663/files#diff-2602188597986ba3e53c5b56162786d2L297
- https://github.com/aspnet/EntityFramework/pull/7663/files#diff-2602188597986ba3e53c5b56162786d2R313
- https://github.com/aspnet/EntityFramework/pull/7663/files#diff-2602188597986ba3e53c5b56162786d2R508
- https://github.com/aspnet/EntityFramework/pull/7663/files#diff-2602188597986ba3e53c5b56162786d2R585
- https://github.com/aspnet/EntityFramework/pull/7663/files#diff-2602188597986ba3e53c5b56162786d2R3115
- https://github.com/aspnet/EntityFramework/pull/7663/files#diff-2602188597986ba3e53c5b56162786d2R3180
- https://github.com/aspnet/EntityFramework/pull/7663/files#diff-2602188597986ba3e53c5b56162786d2R5521
- https://github.com/aspnet/EntityFramework/pull/7663/files#diff-2602188597986ba3e53c5b56162786d2R6207

File: src/Microsoft.EntityFrameworkCore.Specification.Tests/ComplexNavigationsQueryTestBase.cs [hunk 1]
```diff
@@ -3202,7 +3202,7 @@ public virtual void GroupJoin_in_subquery_with_client_projection()
                             where (from l1_inner in context.LevelOne.ToList()
                                    join l2_inner in context.LevelTwo.ToList() on l1_inner.Id equals l2_inner.Level1_Optional_Id into grouping
                                    from l2_inner in grouping.DefaultIfEmpty()
-                                   select l1_inner).Distinct().Count() > 7
+                                   select ClientStringMethod(l1_inner.Name)).Count() > 7
```
Comment (maumar) [inline src/Microsoft.EntityFrameworkCore.Specification.Tests/ComplexNavigationsQueryTestBase.cs]: whoops :)

File: src/Microsoft.EntityFrameworkCore.Relational/Query/RelationalQueryModelVisitor.cs [hunk 2]
```diff
@@ -1401,6 +1559,75 @@ var discriminatorPredicate
...
+                var ancestor = _queryModelVisitor.ParentQueryModelVisitor;
+
+                if (ancestor != null && !ancestor.CanBindOuterProperties)
+                {
+                    return false;
+                }
+
+                do
+                {
+                    if (ancestor.TryGetQuery(outerPropertyExpression.QuerySource) != null)
```
Comment (maumar) [inline src/Microsoft.EntityFrameworkCore.Relational/Query/RelationalQueryModelVisitor.cs]: NRE

File: src/Microsoft.EntityFrameworkCore.Relational/Extensions/RelationalExpressionExtensions.cs [hunk 3]
```diff
@@ -34,5 +34,17 @@ public static bool IsSimpleExpression([NotNull] this Expression expression)
                    || unwrappedExpression is ParameterExpression
                    || unwrappedExpression.IsAliasWithColumnExpression();
         }
+
+        public static Expression MaybeAnonymousSubquery([CanBeNull] this Expression expression)
```
Comment (smitpatel) [inline src/Microsoft.EntityFrameworkCore.Relational/Extensions/RelationalExpressionExtensions.cs]: Can you elaborate what this would generate and why it is needed?

File: src/Microsoft.EntityFrameworkCore.Relational/Query/ExpressionVisitors/ISqlTranslatingExpressionVisitorFactory.cs [hunk 4]
```diff
@@ -16,18 +15,16 @@ public interface ISqlTranslatingExpressionVisitorFactory
...
         ///     A SqlTranslatingExpressionVisitor.
         /// </returns>
         SqlTranslatingExpressionVisitor Create(
             [NotNull] RelationalQueryModelVisitor queryModelVisitor,
-            [CanBeNull] SelectExpression targetSelectExpression = null,
             [CanBeNull] Expression topLevelPredicate = null,
-            bool bindParentQueries = false,
-            bool inProjection = false);
+            bool inProjection = false,
+            bool mutateProjections = true);
```
Comment (smitpatel) [inline src/Microsoft.EntityFrameworkCore.Relational/Query/ExpressionVisitors/ISqlTranslatingExpressionVisitorFactory.cs]: Purpose of `mutateProjections`?

File: src/Microsoft.EntityFrameworkCore.Relational/Query/RelationalQueryModelVisitor.cs (hunk 2 above)
Comment (tuespetre) [inline src/Microsoft.EntityFrameworkCore.Relational/Query/RelationalQueryModelVisitor.cs]: thanks, will change it from `do...while` to `while`

File: src/Microsoft.EntityFrameworkCore.Relational/Extensions/RelationalExpressionExtensions.cs (hunk 3 above)
Comment (tuespetre) [inline src/Microsoft.EntityFrameworkCore.Relational/Extensions/RelationalExpressionExtensions.cs]: `SqlTranslatingExpressionVisitor.VisitSubQuery` could return a `SelectExpression`, but it was always setting the `Alias` to `string.Empty`, which can cause invalid SQL in some cases now that more queries can be inlined. The solution was to only make the resulting `SelectExpression` into an anonymous subquery when appropriate.

Comment (smitpatel) [inline src/Microsoft.EntityFrameworkCore.Relational/Extensions/RelationalExpressionExtensions.cs]: Ah, I see. We need to investigate why it is needed. Ideally, we start with top level select expression without setting any alias. We need alias only when we add it as subquery in another select expression. While flattening can cause one of more select expression to be joined, there is no code path which would cause select expression to split, hence we should not have to set the alias to `string.Empty` ever. If we are doing so then we are assigning alias incorrectly in first place which should be fixed.

File: test/Microsoft.EntityFrameworkCore.SqlServer.FunctionalTests/ComplexNavigationsQuerySqlServerTest.cs [hunk 5]
```diff
@@ -1706,13 +1692,10 @@ public override void GroupJoin_on_complex_subquery_and_set_operation_on_grouping
...
-    FROM [Level1] AS [l10]
-    LEFT JOIN [Level2] AS [l1.OneToOne_Required_FK] ON [l10].[Id] = [l1.OneToOne_Required_FK].[Level1_Required_Id]
-    WHERE ([l10].[Name] <> N'L1 01') OR [l10].[Name] IS NULL
-) AS [t]",
+                @"SELECT [l10].[Id], [l10].[Date], [l10].[Name], [l10].[OneToMany_Optional_Self_InverseId], [l10].[OneToMany_Required_Self_InverseId], [l10].[OneToOne_Optional_SelfId], [l1.OneToOne_Required_FK].[Id], [l1.OneToOne_Required_FK].[Date], [l1.OneToOne_Required_FK].[Level1_Optional_Id], [l1.OneToOne_Required_FK].[Level1_Required_Id], [l1.OneToOne_Required_FK].[Name], [l1.OneToOne_Required_FK].[OneToMany_Optional_InverseId], [l1.OneToOne_Required_FK].[OneToMany_Optional_Self_InverseId], [l1.OneToOne_Required_FK].[OneToMany_Required_InverseId], [l1.OneToOne_Required_FK].[OneToMany_Required_Self_InverseId], [l1.OneToOne_Required_FK].[OneToOne_Optional_PK_InverseId], [l1.OneToOne_Required_FK].[OneToOne_Optional_SelfId]
```
Comment (maumar) [inline test/Microsoft.EntityFrameworkCore.SqlServer.FunctionalTests/ComplexNavigationsQuerySqlServerTest.cs]: I think I have a relatively simple fix in progress that makes this better, so that we don't regress here

File: src/Microsoft.EntityFrameworkCore.Relational/Extensions/RelationalExpressionExtensions.cs (hunk 3 above)
Comment (tuespetre) [inline src/Microsoft.EntityFrameworkCore.Relational/Extensions/RelationalExpressionExtensions.cs]: Well the string.Empty makes it a parenthesized, anonymous subquery. It's just that with new optimizations, it should only be applied on the right occasion instead of unconditionally like it was before.

where (select count(*) from ....) > 7

and so forth. Sorry for the formatting, on a phone right now

File: test/Microsoft.EntityFrameworkCore.SqlServer.FunctionalTests/ComplexNavigationsQuerySqlServerTest.cs (hunk 5 above)
Comment (tuespetre) [inline test/Microsoft.EntityFrameworkCore.SqlServer.FunctionalTests/ComplexNavigationsQuerySqlServerTest.cs]: Is this a regression though? The resulting SQL is smaller

File: src/Microsoft.EntityFrameworkCore.Relational/Query/ExpressionVisitors/ISqlTranslatingExpressionVisitorFactory.cs (hunk 4 above)
Comment (tuespetre) [inline src/Microsoft.EntityFrameworkCore.Relational/Query/ExpressionVisitors/ISqlTranslatingExpressionVisitorFactory.cs]: There are places where the visitor should not be adding to the projections, like in the projection visitor, some result operator handlers... this prevents every place from having to use a 'projection count map' or whatever

File: test/Microsoft.EntityFrameworkCore.SqlServer.FunctionalTests/ComplexNavigationsQuerySqlServerTest.cs (hunk 5 above)
Comment (smitpatel) [inline test/Microsoft.EntityFrameworkCore.SqlServer.FunctionalTests/ComplexNavigationsQuerySqlServerTest.cs]: @maumar - Subquery is getting lifted. Better SQL.

File: src/Microsoft.EntityFrameworkCore.Relational/Query/ExpressionVisitors/ISqlTranslatingExpressionVisitorFactory.cs (hunk 4 above)
Comment (smitpatel) [inline src/Microsoft.EntityFrameworkCore.Relational/Query/ExpressionVisitors/ISqlTranslatingExpressionVisitorFactory.cs]: Can we name it something different? We are trying to make SelectExpression as less `Mutable` as possible. 😄

File: src/Microsoft.EntityFrameworkCore.Relational/Extensions/RelationalExpressionExtensions.cs (hunk 3 above)
Comment (smitpatel) [inline src/Microsoft.EntityFrameworkCore.Relational/Extensions/RelationalExpressionExtensions.cs]: Good point. Are all such calls go through single path where we can put it instead of having a function to do it?

File: src/Microsoft.EntityFrameworkCore.Relational/Query/ExpressionVisitors/ISqlTranslatingExpressionVisitorFactory.cs (hunk 4 above)
Comment (tuespetre) [inline src/Microsoft.EntityFrameworkCore.Relational/Query/ExpressionVisitors/ISqlTranslatingExpressionVisitorFactory.cs]: Sure!

File: src/Microsoft.EntityFrameworkCore.Relational/Extensions/RelationalExpressionExtensions.cs (hunk 3 above)
Comment (tuespetre) [inline src/Microsoft.EntityFrameworkCore.Relational/Extensions/RelationalExpressionExtensions.cs]: They did go through a single path but now not everything coming out of that single path should have it applied

File: test/Microsoft.EntityFrameworkCore.SqlServer.FunctionalTests/ComplexNavigationsQuerySqlServerTest.cs (hunk 5 above)
Comment (tuespetre) [inline test/Microsoft.EntityFrameworkCore.SqlServer.FunctionalTests/ComplexNavigationsQuerySqlServerTest.cs]: Oh it looks like the selected columns are different. (More being selected.)

**Comment by tuespetre** [comment]
:up::date:

- Fixed the possible NRE pointed out
- Renamed 'mutateProjections' parameter to 'addToProjections'
- Undid the changes related to #7619

File: src/Microsoft.EntityFrameworkCore.Relational/Query/ExpressionVisitors/RelationalProjectionExpressionVisitor.cs [hunk 6]
```diff
@@ -63,23 +63,65 @@ protected override Expression VisitMethodCall(MethodCallExpression node)
...
+                    node,
+                    node.Expression.RemoveConvert(),
+                    node.Member.Name);
+
+            if (handledPropertyExpression != null)
+            {
+                return handledPropertyExpression;
+            }
+
+            var handledMemberExpression
```
Comment (smitpatel) [inline src/Microsoft.EntityFrameworkCore.Relational/Query/ExpressionVisitors/RelationalProjectionExpressionVisitor.cs]: Use null propagation.

File: src/Microsoft.EntityFrameworkCore.Relational/Query/ExpressionVisitors/RelationalProjectionExpressionVisitor.cs [hunk 7]
```diff
@@ -63,23 +63,65 @@ protected override Expression VisitMethodCall(MethodCallExpression node)
 
             if (EntityQueryModelVisitor.IsPropertyMethod(node.Method))
             {
-                var newArg0 = Visit(node.Arguments[0]);
+                var handledPropertyExpression
+                    = TryHandlePropertyExpression(
+                        node,
+                        node.Arguments.FirstOrDefault().RemoveConvert(),
```
Comment (smitpatel) [inline src/Microsoft.EntityFrameworkCore.Relational/Query/ExpressionVisitors/RelationalProjectionExpressionVisitor.cs]: Possible NRE.
Also why we need to use FirstOrDefault instead of direct indexes? Property method always have 2 args.

File: src/Microsoft.EntityFrameworkCore.Relational/Extensions/RelationalExpressionExtensions.cs (hunk 3 above)
Comment (smitpatel) [inline src/Microsoft.EntityFrameworkCore.Relational/Extensions/RelationalExpressionExtensions.cs]: I looked into usages of this function. They are being used at various expressions without any corresponding logic around it. We shouldn't have it this way.
Encapsulate this code into the particular code path with condition where this is required. Subquery with empty string as alias is pitfall for many invalid SQL. We should identify the pattern where this is required and apply for those cases only. Ideally this function should be removed.

File: src/Microsoft.EntityFrameworkCore/Query/QueryCompilationContext.cs [hunk 8]
```diff
@@ -30,6 +30,8 @@ public class QueryCompilationContext
     {
         private readonly IRequiresMaterializationExpressionVisitorFactory _requiresMaterializationExpressionVisitorFactory;
         private readonly IEntityQueryModelVisitorFactory _entityQueryModelVisitorFactory;
+        private readonly IDictionary<QueryModel, EntityQueryModelVisitor> _entityQueryModelVisitors
```
Comment (smitpatel) [inline src/Microsoft.EntityFrameworkCore/Query/QueryCompilationContext.cs]: 👍

Comment (smitpatel) [inline src/Microsoft.EntityFrameworkCore/Query/QueryCompilationContext.cs]: This change is good in a way that we have reference to QMV which visited this QM already. Hence avoiding revising same QM again. @anpete - Any side effects of this? (like visiting a QM adding something to SelectExpression of some other QM)

File: src/Microsoft.EntityFrameworkCore.Relational/Query/ExpressionVisitors/RelationalProjectionExpressionVisitor.cs (hunk 6 above)
Comment (tuespetre) [inline src/Microsoft.EntityFrameworkCore.Relational/Query/ExpressionVisitors/RelationalProjectionExpressionVisitor.cs]: ✔️

File: src/Microsoft.EntityFrameworkCore.Relational/Query/ExpressionVisitors/RelationalProjectionExpressionVisitor.cs (hunk 7 above)
Comment (tuespetre) [inline src/Microsoft.EntityFrameworkCore.Relational/Query/ExpressionVisitors/RelationalProjectionExpressionVisitor.cs]: We don't need to, that should be an indexer ✔️

File: src/Microsoft.EntityFrameworkCore.Relational/Query/ExpressionVisitors/RelationalProjectionExpressionVisitor.cs [hunk 9]
```diff
@@ -90,11 +132,32 @@ protected override Expression VisitNew(NewExpression expression)
         {
             Check.NotNull(expression, nameof(expression));
 
-            var newNewExpression = base.VisitNew(expression);
+            if (QueryModelVisitor.QueryCompilationContext.QuerySourceRequiresMaterialization(_querySource))
```
Comment (tuespetre) [inline src/Microsoft.EntityFrameworkCore.Relational/Query/ExpressionVisitors/RelationalProjectionExpressionVisitor.cs]: @smitpatel, regarding `MaybeAnonymousSubquery()`:

I added this condition as a workaround for not being able to always inline a subquery into a projection (due to an `EntityShaper` being present, or otherwise.) It looks like maybe this takes care of the reason I had put in `MaybeAnonymousSubquery` so I can remove it after all. 🎉

File: src/Microsoft.EntityFrameworkCore.Relational/Query/RelationalQueryModelVisitor.cs (hunk 2 above)
Comment (tuespetre) [inline src/Microsoft.EntityFrameworkCore.Relational/Query/RelationalQueryModelVisitor.cs]: actually changed the preceding condition to return early if the ancestor is null instead

**Comment by tuespetre** [comment]
:up::date:

Was able to remove `MaybeAnonymousSubquery()`

File: src/Microsoft.EntityFrameworkCore/Query/QueryCompilationContext.cs (hunk 8 above)
Comment (smitpatel) [inline src/Microsoft.EntityFrameworkCore/Query/QueryCompilationContext.cs]: @tuespetre - This change would improve perf for query compiler though there is a slight chance of side-effects of caching querymodelvisitors if there are any changed flags in parent qmv in between multiple visits of same QM. The only way to verify is to just make caching changes and see if no tests change. Can you submit a separate PR for that?

Comment (tuespetre) [inline src/Microsoft.EntityFrameworkCore/Query/QueryCompilationContext.cs]: I have approached it that way before (it was one of the first split-out PRs I tried to do) but it does not work. :chicken: ➕ 🥚; also with these changes there is no multiple visiting of the same query model.

Comment (smitpatel) [inline src/Microsoft.EntityFrameworkCore/Query/QueryCompilationContext.cs]: If caching QMVs is breaking tests that means it has unintentional side-effects and we cannot cache it. While binding/lifting, QMV peeks into parent QMV and even though same query model is being visited it can produce different results. Mostly, caching is not usable here.

Comment (tuespetre) [inline src/Microsoft.EntityFrameworkCore/Query/QueryCompilationContext.cs]: > even though same query model is being visited it can produce different results

That's my point, there same query model will not be visited more than once

Comment (tuespetre) [inline src/Microsoft.EntityFrameworkCore/Query/QueryCompilationContext.cs]: > If caching QMVs is breaking tests that means it has unintentional side-effects

The problem I found was not with the caching itself, but that introducing caching actually **prevents** the query models from being visited more than once, and once you do that, you have to use/account for new and different flags on the parent QMV, such as `CanBindOuterProperties`, `RequiresOuterParameterBinding`, etc.

Comment (tuespetre) [inline src/Microsoft.EntityFrameworkCore/Query/QueryCompilationContext.cs]: I will try removing the from this PR and see if it breaks anything. If it doesn't break, I will push up a commit that doesn't have the caching. Does that sound good to you

Comment (smitpatel) [inline src/Microsoft.EntityFrameworkCore/Query/QueryCompilationContext.cs]: I tested out caching part and it breaks many tests. In current pipeline, visiting a QM not only dependent on outer qmv flags but also mutates the QM itself. When we fail to translate the QM, we have to revert the QM. We have methods like `PopulateQueryModelMapping` & `RecreateQueryModelFromMapping` for that. That is enough evidence to say that current pipeline generate different results while visiting same query model again.

Before we introduce caching on QMVs, we need to make sure that pipeline can generate the required translation of QM which can be used under any circumstances later.

Remove the caching part from this PR. We can add it when we reach above stage.

Comment (tuespetre) [inline src/Microsoft.EntityFrameworkCore/Query/QueryCompilationContext.cs]: Ohhh you gave me an idea for how to deal with all the unnecessary uniquification without the caching bits (for another PR, another time) and also I think I know how to get around the removal of the caching for this PR.
